cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java

```diff
@@ -0,0 +1,229 @@
+      }
+      if (!entry.isPayloadReference()) {
+        putBuilder.add(columnFamily, PAYLOAD_COL, entry.getPayload());
+      }
+      batchPuts.add(putBuilder.build());
+    }
+    if (!batchPuts.isEmpty()) {
+      hTable.put(batchPuts);
+      if (!hTable.isAutoFlush()) {
+        hTable.flushCommits();
+      }
+    }
+  }
+
+  @Override
+  public void delete(TopicId topicId, long transactionWritePointer) throws IOException {
+    byte[] rowPrefix = MessagingUtils.toBytes(topicId);
+    byte[] stopRow = Bytes.stopKeyForPrefix(rowPrefix);
+    // TODO: Optimize this scan by passing in a messageId since this scan can be
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(rowPrefix)
+      .setStopRow(stopRow)
+      .build();
+
+    List<Delete> batchDeletes = new ArrayList<>();
+    try (ResultScanner scanner = hTable.getScanner(scan)) {
+      Iterator<Result> iterator = scanner.iterator();
+      Result result = iterator.next();
+      if (result.containsColumn(columnFamily, TX_COL)) {
+        byte[] txCol = result.getValue(columnFamily, TX_COL);
+        if (Bytes.equals(txCol, Bytes.toBytes(transactionWritePointer))) {
+          Delete delete = tableUtil.buildDelete(result.getRow())
+            .build();
+          batchDeletes.add(delete);
+        }
+      }
+    }
+
+    if (!batchDeletes.isEmpty()) {
+      hTable.delete(batchDeletes);
+      if (!hTable.isAutoFlush()) {
+        hTable.flushCommits();
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (hTable != null) {
+      hTable.close();
+    }
+  }
+
+  private static class HBaseCloseableIterator extends AbstractCloseableIterator<Entry> {
+    private final ResultScanner scanner;
+    private final Transaction transaction;
+    private final byte[] colFamily;
+    private int limit;
+    private boolean closed = false;
+
+    HBaseCloseableIterator(ResultScanner scanner, int limit, byte[] colFamily, @Nullable Transaction transaction) {
+      this.scanner = scanner;
+      this.transaction = transaction;
+      this.colFamily = colFamily;
+      this.limit = limit;
+    }
+
+    @Override
+    protected Entry computeNext() {
+      if (closed) {
+        return endOfData();
+      }
+
+      if (limit <= 0) {
+        return endOfData();
+      }
+
+      Result result = null;
+      try {
+        result = scanner.next();
+      } catch (IOException e) {
+        Throwables.propagate(e);
+      }
+
+      if (result != null) {
+        MessageTable.Entry entry = new DefaultMessageTableEntry(
+          result.getRow(), result.getValue(colFamily, PAYLOAD_COL), result.getValue(colFamily, TX_COL));
+        // If it is a valid entry, return it. Else close the iterator (to guarantee ordering).
+        if (isVisible(entry, transaction)) {
+          limit--;
+          return entry;
+        }
+      }
+      return endOfData();
+    }
+
+    @Override
+    public void close() {
+      scanner.close();
```

Comment:
you might want to wrap this with `try` and have `endOfData(); closed = true;` in finally block.